crates/cpuid-utils/src/host.rs

```diff
@@ -0,0 +1,225 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use bhyve_api::{VmmCtlFd, VmmFd};
+use propolis_types::{CpuidIdent, CpuidValues, CpuidVendor};
+use thiserror::Error;
+
+use crate::{
+    bits::{
+        AmdExtLeaf1DCacheType, AmdExtLeaf1DEax, Leaf1Ecx, EXTENDED_BASE_LEAF,
+        STANDARD_BASE_LEAF,
+    },
+    CpuidSet, SubleafInsertConflict,
+};
+
+#[derive(Debug, Error)]
+pub enum GetHostCpuidError {
+    #[error("failed to insert into the CPUID map")]
+    CpuidInsertFailed(#[from] SubleafInsertConflict),
+
+    #[error("CPUID vendor not recognized: {0}")]
+    VendorNotRecognized(&'static str),
+
+    #[error("I/O error from bhyve API")]
+    BhyveError(#[from] std::io::Error),
+}
+
+/// A wrapper around a handle to a bhyve VM that can be used to query bhyve's
+/// default CPUID values.
+struct Vm(bhyve_api::VmmFd);
+
+impl Vm {
+    fn new() -> Result<Self, GetHostCpuidError> {
+        let name = format!("cpuid-gen-{}", std::process::id());
+        let mut req = bhyve_api::vm_create_req::new(name.as_bytes())
+            .expect("valid VM name");
+
+        let ctl = VmmCtlFd::open()?;
+        let _ = unsafe { ctl.ioctl(bhyve_api::VMM_CREATE_VM, &mut req) }?;
+
+        let vm = match VmmFd::open(&name) {
+            Ok(vm) => vm,
+            Err(e) => {
+                // Attempt to manually destroy the VM if we cannot open it
+                let _ = ctl.vm_destroy(name.as_bytes());
+                return Err(e.into());
+            }
+        };
+
+        Ok(Self(vm))
+    }
+
+    fn query(
+        &self,
+        vlc_eax: u32,
+        vlc_ecx: u32,
+    ) -> Result<CpuidValues, GetHostCpuidError> {
+        let mut data = bhyve_api::vm_legacy_cpuid {
+            vlc_eax,
+            vlc_ecx,
+            ..Default::default()
+        };
+        unsafe { self.0.ioctl(bhyve_api::VM_LEGACY_CPUID, &mut data) }?;
+        Ok(CpuidValues {
+            eax: data.vlc_eax,
+            ebx: data.vlc_ebx,
+            ecx: data.vlc_ecx,
+            edx: data.vlc_edx,
+        })
+    }
+}
+
+impl Drop for Vm {
+    fn drop(&mut self) {
+        let _ = self.0.ioctl_usize(bhyve_api::VM_DESTROY_SELF, 0);
+    }
+}
+
+/// Queries the supplied CPUID leaf on the caller's machine.
+#[cfg(target_arch = "x86_64")]
```

Comment:
nit: could use x86 here, rather than demanding x86-64, but it's not the first thing the cpuid area to be like that

Reply:
I gave this a try but discovered that `rustc` appears to have no 32-bit illumos target! 

```
$ rustc --version
rustc 1.84.0 (9fc6b4312 2025-01-07)
$ rustc --print=target-list | grep illumos
aarch64-unknown-illumos
x86_64-unknown-illumos
```

I don't think we're actually querying host CPUID in any platform-agnostic tests that one might want to run against an i686 target, so I left this as-is. (I think supporting x86 would also require defining these functions twice, since the CPUID intrinsic is in `core::arch::x86` when compiling against that target, unless there's some subtlety that I'm missing.)